### PR TITLE
add higher level eth package implementing much of eip-4844 bytes api

### PR DESCRIPTION
**Pull request by roberto-bayardo**
Hi @protolambda , two things in particular I need some guidance on:

Should the versioned hash stuff be in here? If not, where do you think it should go?  (It's used both by CL and EL clients.)

I'm depending on go-ethereum/params ... should I just copy  / paste those needed constants here to remove that dependency?

thanks.

### CODE REVIEW

**Comment by roberto-bayardo** [comment]
Oh and I will push tests in a separate PR, as they are unfortunately are not yet quite ready to pull out of geth.

**Comment by protolambda** [comment]
> I'm depending on go-ethereum/params ... should I just copy / paste those needed constants here to remove that dependency?

Yes please

**Comment by protolambda** [comment]
> Should the versioned hash stuff be in here? If not, where do you think it should go? (It's used both by CL and EL clients.)

For completeness I think it's fine to host here. The complexity is low, and it may come in useful once we have standard tests for the eth KZG interface usage, to be able to run them all.

**Comment by roberto-bayardo** [comment]
> 

done

**Comment by roberto-bayardo** [comment]
cc: @kevaundray

File: eth/helpers.go [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+	SetupG2       []bls.G2Point `json:"setup_G2"`
+	SetupLagrange []bls.G1Point `json:"setup_G1_lagrange"`
+}
+
+// Initialize KZG subsystem (load the trusted setup data)
+func init() {
+	var parsedSetup = JSONTrustedSetup{}
+
+	// TODO: This is dirty. KZG setup should be loaded using an actual config file directive
+	err := json.Unmarshal([]byte(KZGSetupStr), &parsedSetup)
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply eth/helpers.go]: moved to trusted_setup.json and using embed to put in the Go var as suggested.

**Comment by protolambda** [comment]
Updating go-kzg min Go requirement to Go 1.18 for `embed` and other newer features to work. See https://github.com/protolambda/go-kzg/pull/29

**Comment by protolambda** [comment]
@roberto-bayardo I updated `master` branch with Go 1.18 from #29. Can you rebase this PR?

Also, can I help update `TxPeekBlobVersionedHashes` to not use `ztyp`? I think it can be more efficient/minimal, and it'll remove a dependency.

**Comment by roberto-bayardo** [comment]
> @roberto-bayardo I updated `master` branch with Go 1.18 from #29. Can you rebase this PR?

done

> Also, can I help update `TxPeekBlobVersionedHashes` to not use `ztyp`? I think it can be more efficient/minimal, and it'll remove a dependency.

yes i'm all for it.

**Comment by protolambda** [comment]
- For some reason the `kzg.go` file was dropped in the rebase, I edited the commit to put it back.
- Updated the `TxPeekBlobVersionedHashes` to explain the offsets better, and do the more minimal thing without external library.

File: bls/bignum_hol256.go [hunk 2]
```diff
@@ -93,7 +93,7 @@ func RandomFr() *Fr {
 }
 
 func SubModFr(dst *Fr, a, b *Fr) {
-	if (*u256.Int)(dst).SubOverflow((*u256.Int)(a), (*u256.Int)(b)) {
+	if _, underflow := (*u256.Int)(dst).SubOverflow((*u256.Int)(a), (*u256.Int)(b)); underflow {
```
Comment (roberto-bayardo) [inline bls/bignum_hol256.go:96]: confirming you meant to include this in this PR?

Comment (roberto-bayardo) [reply bls/bignum_hol256.go:96]: oh I see you've referenced fixing the build in the commit title.

**Comment by roberto-bayardo** [comment]
> * For some reason the `kzg.go` file was dropped in the rebase, I edited the commit to put it back.

Sorry my bad I had renamed kzg.go from the new eth package and must have got this confused with it.

> * Updated the `TxPeekBlobVersionedHashes` to explain the offsets better, and do the more minimal thing without external library.

LGTM

File: eth/eth.go [hunk 3]
```diff
@@ -209,53 +207,46 @@ func ValidateBlobsSidecar(slot Slot, beaconBlockRoot Root, expectedKZGCommitment
...
-
-	// Advance dr to the versioned hash list. We subtract blobMessageLen from the offset here to
-	// account for the fact that the offset is relative to the position of "message" (70) and we
-	// are currently positioned at the end of it (262).
-	skip := uint64(offset) - blobMessageLen
-	skipped, err := dr.Skip(skip)
-	if err != nil {
-		return nil, fmt.Errorf("could not skip to versioned hashes: %v", err)
+	offset := binary.LittleEndian.Uint32(tx[BlobVersionedHashesOffset : BlobVersionedHashesOffset+4])
+	if uint64(offset) > uint64(len(tx)) {
```
Comment (roberto-bayardo) [inline eth/eth.go:242]: is this right? offset is relative to tx[70] not tx[0] I believe, so perhaps you just need to add 70 to it first.

Comment (protolambda) [reply eth/eth.go:242]: nice catch, will fix

**Comment by roberto-bayardo** [comment]
cool looks good to go? Once merged I'll run it through our interop repo e2e tests.

File: eth/eth.go (hunk 3 above)
Comment (protolambda) [reply eth/eth.go]: updated, ready for review. I think we need some standard json tests for this, like we did with BLS: https://github.com/ethereum/bls12-381-tests/ then we can catch it before plugging it into a client (with its own testing that should cover this).

Comment (roberto-bayardo) [reply eth/eth.go]: lg, when i move the tests still in the geth fork over I can add something for this function.

File: eth/domain.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	primitiveRoot := big.NewInt(7)
+	width := big.NewInt(int64(FieldElementsPerBlob))
+	exp := new(big.Int).Div(new(big.Int).Sub(BLSModulus, big.NewInt(1)), width)
+	rootOfUnity := new(big.Int).Exp(primitiveRoot, exp, BLSModulus)
+	DomainFr = make([]bls.Fr, FieldElementsPerBlob)
+	for i := 0; i < FieldElementsPerBlob; i++ {
+		// We reverse the bits of the index as specified in https://github.com/ethereum/consensus-specs/pull/3011
+		// This effectively permutes the order of the elements in Domain
+		reversedIndex := reverseBits(uint64(i), FieldElementsPerBlob)
+		domain := new(big.Int).Exp(rootOfUnity, big.NewInt(int64(reversedIndex)), BLSModulus)
```
Comment (protolambda) [inline eth/domain.go:31]: FWIW, I think we can speed this up by using a temporary variable for the output, multiplying it by the root of unity each iteration, and copying that into the output, and then shuffling the output to be in reverse bit order afterwards. But this works fine too.

Comment (roberto-bayardo) [reply eth/domain.go:31]: thx will keep in mind for the next iteration

**Pull request merged**